### ISSUE TITLE
Scaling up Test Discovery for large solutions

### ISSUE THREAD

**Issue by sbaid** [Microsoft/vstest#674]
## Description
Test Explorer discovers _all_ tests containers* and tests in the _changed_ test containers when the solution is opened or is built. This poses following issues for large solutions
1. On Solution load Test discovery for a solution containing 50+ projects & order of 10K+ tests takes order of minutes. User need to wait for it. It is neither optional nor cancelable.
2. On solution build - all _changed_ containers are discovered. User cannot control what projects should be discovered.

*Test Container could be a .dll for managed code, .js for javascript tests.

### Proposed solution
- [ ] 1. Give user control over the discovery operation - Make solution load time (auto) discovery optional through a Tools->option and make it cancelable. Provide an explicit discover tests menu besides run all.
- [ ] 2. Discover only the test containers filtered by _Project_ and then discover tests in those test containers. On Build, rediscover tests in only the filtered test projects.
![image](https://cloud.githubusercontent.com/assets/19462879/24541240/3e7518ee-1614-11e7-8ebf-1e0af665a1be.png)

**Comment by sbaid** [comment]
## Test Container Discovery
### Background
Test Window allows ITestContainerDiscoverer based extensions to report test containers. Test Containers can be .dll,  .appx, .js file etc which contain tests. These containers are discovered at solution load time and when ITestContainerDiscoverer.TestContainersUpdated event is fired. Thereafter  test discovery is done in the changed test containers by using the test adapters.

Test window implements VsProjectOutputContainerDiscoverer that discovers all projects in the solution, queries their outputpath and reports test containers.

Test container discovery currently has following issues. 
	1. There is only one ITestContainerDiscoverer implementation in Test Window code. VsProjectOutputContainerDiscoverer handles a) CPS b) NON-CPS c) UWP d) native project systems. 
	2. ITestContainerDiscoverer interface does not support async query of test containers & does not support query of test containers filtered by a pattern.
	3. The implementation uses IVsSolution.GetProjectEnum() and is invoked at solution load time & after each build. This call may be costly. The current code uses __VSENUMPROJFLAGS.EPF_LOADEDINSOLUTION. and hence, does not enumerate projects not-loaded projects in DPL mode.

### Solution
#### Overview
Have a different ITestContainerDiscoverer impl for each kind of runtime/appmodel 
	1. NetCoreTestContainerDiscoverer
	2. NetFxTestContainerDiscoverer
	3. UWPTestContainerDiscoverer
	4. NativeTestContainerDiscoverer
Once this is done, 
1. Each discoverer can optimize how to enumerate its test containers. e.g. NetCore test projects have "testcontainer" capability.  NativeCpp test projects can probably use 
<ProjectSubType>NativeUnitTestProject</ProjectSubType>
To filter test projects.
2. These TestContainerDiscoverers need to work in DPL mode & non-DPL mode. 
3. These test containers need to optimize & report incremental changes in test containers when they are queried subsequently (after initial discovery).

#### Details
```CSharp
ITestContainerDiscoverer2
{
    Uri ExecutorUri { get; }
    Task<IEnumerable<TestContainerResult>> QueryAsync(string sourceNamePattern, CancellationToken token);
    event EventHandler TestContainersUpdated;
}

Class TestContainerResult 
{
    ITestContainer TestContainer { get;  private set; }
    TestContainerChange  { get; private set; }
}

enum TestContainerChange
{
    Invalid = 0, Add = 1, Remove = 2, Updated = 3
}
```
### Discussion 
1. What api to use for the project enumeration?
		a. Will the api be same when DPL mode is enabled or disabled?
		b. Is IVsSolution.GetProjectEnum() the right api?
		c. How to add project name based filter when enumerating the projects? 
2.  IVSSolution.GetProjectEnum() is currently used to enumerate projects in a solution. What API should the test container discoverer for CPS/Non-CPS project system use and should they use the same api when under DPL or non-DPL mode?
3. How to query project output path in DPL mode? is there one way of query it in DPL/NON-DPL mode?

Contributions will be welcome to build test container discoverers. Also in Test Explorer codebase (for Msft) as it is not yet open source.

**Comment by sbaid** [comment]
@viktorveis @olgaark @srivatsn @davkean @mmanela

**Comment by mmanela** [comment]
What about file based extensions like Chutzpah?

**Comment by olgaark** [comment]
Probably the most performant way is to make all test projects (or files) to somehow register with test toolwindow  instead of it asking all projects in the solution. 

If this does not sound like a feasible/extendable idea, checking for a project capability (and requiring it to be set in all test projects) would be the next choice - it will work in DPL as well as c++ FPL (project cache) mode without forcing a real project to load.

**Comment by davkean** [comment]
You could do a mix of both; have a component that is only present when the TestContainer capability that registers itself with the tool window.

**Comment by viktorveis** [comment]
Today, lightweight solution load (LSL) only discovers tests from loaded projects. First, we should make test discoverability consistent between LSL and normal solution load. Then we can add a way to filter tests as you proposed. LSL has API to discover all projects in the solution without loading them: https://docs.microsoft.com/en-us/visualstudio/extensibility/lightweight-solution-load-extension-impact

**Comment by sbaid** [comment]
@mmanela Chutzpah (file based test containers) can continue to implement a test container discovery & provide ITestContainerDiscover2 impl once it is available.

**Comment by sbaid** [comment]
@viktorveis I read through documentation. We need your contribution towards enabling LSL into test explorer -- by adding `NetFxLslTestContainerDiscoverer` extension.

Here is my notes.
1. LSL is supported by NetFx (Full CLR) projects but _not_ by Native, UWP, NET core in 15.3 timeframe.
2. `NetFxLslTestContainerDiscoverer` will implement `ITestContainerDiscoverer` interface.

```Csharp
public interface ITestContainerDiscoverer
    {
        /// <summary>
        /// The Uri for the test adapter that handles test from containers from this discoverer
        /// </summary>
        Uri ExecutorUri { get; }

        /// <summary>
        /// Gets test containers which will be used as the source to discover individual tests
        /// </summary>
        /// <returns>Collection of ITestContainer which holds the test container and debug engine ID(s)</returns>
        IEnumerable<ITestContainer> TestContainers { get; }

        /// <summary>
        /// Event that is fired when test containers maybe updated.
        /// (For example on build or save)
        /// </summary>
        event EventHandler TestContainersUpdated;
    }
```
### NetFxLslTestContainerDiscoverer
1. It will return test containers only if the current solution has any deferred load projects.
2.  `TestContainers` property will _enumerate_ deferred projects. I think it will use `GetProjectEnum` with  `__VSENUMPROJFLAGS3.EPF_DEFERRED`
3. File the event `TestContainersUpdated` when the containers are updated.

```CSharp
public interface ITestContainer
    {
        /// <summary>
        /// The <seealso cref="ITestContainerDiscoverer"/> that discoverered the container
        /// </summary>
        ITestContainerDiscoverer Discoverer { get; }

        /// <summary>
        /// Return test container source which allows the test adapter to discover tests within it.
        /// </summary>
        string Source { get; }

        /// <summary>
        /// Returns the debug engine(s) to be used by VisualStudio debugger
        /// </summary>
        IEnumerable<Guid> DebugEngines { get; }

        /// <summary>
        /// Target .net framework of this test container. Use FrameworkVersion.None if this property is not applicable. 
        /// </summary>
        FrameworkVersion TargetFramework { get; }

        /// <summary>
        /// Target platform of this test container. Use Architecture.Default if this property is not applicable. 
        /// </summary>
        Architecture TargetPlatform { get; }

        /// <summary>
        ///True if this is a test container for Windows Store app tests. False otherwise.
        /// </summary>
        bool IsAppContainerTestContainer { get; }

        /// <summary>
        /// Deploys the Windows Store app test container and returns info about the deployment, or does nothing and returns null if this is not a Windows Store app test container.
        /// </summary>
        IDeploymentData DeployAppContainer();

        /// <summary>
        /// Compares with <paramref name="other"/>
        /// </summary>
        int CompareTo(ITestContainer other);

        /// <summary>
        /// Creates an snapshot in time of the container
        /// </summary>
        /// <returns></returns>
        ITestContainer Snapshot();
    }

public interface IVsTestContainer : ITestContainer
    {
        IVsProject Project { get; }
    }
```
1. `Source` property need to query the project output path. It likely need to use `GetManagedCommandLineInfoForConfigurationAsyn` and use `managedcommandLineInfo.TargetPath` property which returns the project output path.

It can be part of `Microsoft.VisualStudio.TestWindow.Extensions.dll` which @rahulpaw  will be adding shortly under `TestWindow\Extensions` folder in vs repo. `VsTestContainer.cs`,  `VsProjectOutputContainerDiscoverer.cs` , `TestContainerProvider.cs` files will be good starting point. 

To enable LSL into test window, we look fwd to your contribution.

**Comment by sbaid** [comment]
### Solution part (1) 

#### Add tools->options->Test
- [ ] Disable test discovery at solution load
This will be off by default. 

For managed projects - another option like this will help.
- [ ] Discover tests only in projects marked with `TestContainer` project capability.
This will be off by default.
Currently, all projects are queried. This will enable querying projects that have this capability. DotNet test project template has this project capability on by default. VS Test Adapters can add this capability in the latest nuget package. For legacy projects, User can add by adding `<ProjectCapability Include="TestContainer" />` in the project file. 

To enable on demand discovery - can be surfaced either by a) `Discover All` button or b) `Discover Tests` menu in the solution explorer on solution node. `Cancel` menu can surface in test explorer same like in case of `Run All`.

#### Tasks
- [ ] Implement Disable on solution load discovery & surface it as a tools option
- [ ] For project output based test containers, Implement discovery only in projects marked with `TestContainer` project capability.
- [ ] Surface Discover All menu
- [ ] Cancelability for the discovery operation.

**Comment by olgaark** [comment]
>>LSL is supported by NetFx (Full CLR) projects but not by Native, UWP, NET core in 15.3 timeframe
The current plan is to support LSL for vc projects in 15.3.

**Comment by srivatsn** [comment]
To me it sounds like discovering tests with TestContainer capability should be on by default so that most people get the perf benefit. I assume the cases where we need to discover tests from projects that are not test projects would be few and for those cases, users can opt-in.

**Comment by sbaid** [comment]
@olgaark Thanks. Also got to know about this from Viktor in our last discussion.
@srivatsn thanks. We need to likely provide this opt-in initially and at later stage as opt-out.  There will be lot of legacy customer test projects using test adapters that would not have updated to add `TestContainer` capability.

**Comment by snkota** [comment]
This is being tracked external to this repo here:
https://developercommunity.visualstudio.com/content/problem/145922/scaling-up-test-discovery-for-large-solutions.html
